core/src/main/java/nu/yona/server/analysis/service/AnalysisEngineService.java

```diff
@@ -73,57 +76,114 @@ private void analyze(ActivityPayload payload, UserAnonymizedDTO userAnonymized,
 	{
-		Date minEndTime = new Date(payload.startTime.getTime() - yonaProperties.getAnalysisService().getConflictInterval());
-		DayActivity dayActivity = getCachedDayActivity(payload, userAnonymized, matchingGoal);
-		Activity activity = dayActivity == null ? null : dayActivity.getLastActivity();
-		if (activity == null || activity.getEndTime().before(minEndTime))
+		DayActivityCacheResult dayActivity = getRegisteredDayActivity(payload, userAnonymized, matchingGoal);
+		if (canCombineWithLastRegisteredActivity(payload, lastRegisteredActivity))
+		{
+			if (isBeyondSkipWindowAfterLastRegisteredActivity(payload, lastRegisteredActivity))
+			{
+				// Update message only if it is within five seconds to avoid unnecessary cache flushes.
+				updateActivityEndTime(payload, userAnonymized, matchingGoal, dayActivity, lastRegisteredActivity);
+			}
+		}
+		else
 		{
 			addActivity(payload, userAnonymized, matchingGoal, dayActivity);
 		}
-		// Update message only if it is within five seconds to avoid unnecessary cache flushes.
-		// Or if the start time is earlier than the existing start time (this can happen with app activity, see below).
-		else if (payload.endTime.getTime() - activity.getEndTime().getTime() >= yonaProperties.getAnalysisService()
-				.getUpdateSkipWindow() || payload.startTime.before(activity.getStartTime()))
+	}
+
+	private boolean isBeyondSkipWindowAfterLastRegisteredActivity(ActivityPayload payload, Activity lastRegisteredActivity)
+	{
+		return payload.endTime.getTime() - lastRegisteredActivity.getEndTime().getTime() >= yonaProperties.getAnalysisService()
+				.getUpdateSkipWindow();
+	}
+
+	private boolean canCombineWithLastRegisteredActivity(ActivityPayload payload, Activity lastRegisteredActivity)
+	{
+		if (lastRegisteredActivity == null)
+		{
+			return false;
+		}
+
+		if (precedesLastRegisteredActivity(payload, lastRegisteredActivity))
 		{
-			updateActivity(payload, userAnonymized, matchingGoal, dayActivity, activity);
+			// This can happen with app activity.
+			// Do not try to combine, add separately.
+			return false;
 		}
+
+		if (isBeyondCombineIntervalWithLastRegisteredActivity(payload, lastRegisteredActivity))
+		{
+			return false;
+		}
+
+		return true;
 	}
 
-	private DayActivity getCachedDayActivity(ActivityPayload payload, UserAnonymizedDTO userAnonymized, Goal matchingGoal)
+	private boolean precedesLastRegisteredActivity(ActivityPayload payload, Activity lastRegisteredActivity)
 	{
-		DayActivity dayActivity = cacheService.fetchDayActivityForUser(userAnonymized.getID(), matchingGoal.getID());
-		if (dayActivity != null && dayActivity.getStartTime().isBefore(getStartOfDay(payload.startTime, userAnonymized)))
+		return payload.startTime.before(lastRegisteredActivity.getStartTime());
+	}
+
+	private boolean isBeyondCombineIntervalWithLastRegisteredActivity(ActivityPayload payload, Activity lastRegisteredActivity)
+	{
+		Date intervalEndTime = new Date(
```

Comment:
It prevents using the old ``Date`` class and it's more explicit about what happens.